pkg/restore/split_client.go

```diff
@@ -242,17 +242,19 @@ func (c *pdClient) sendSplitRegionRequest(
 	ctx context.Context, regionInfo *RegionInfo, keys [][]byte,
 ) (*kvrpcpb.SplitRegionResponse, error) {
 	var splitErrors error
-	for i := 0; i < splitRegionMaxRetryTime; i++ {
-		var peer *metapb.Peer
-		if regionInfo.Leader != nil {
-			peer = regionInfo.Leader
-		} else {
-			if len(regionInfo.Region.Peers) == 0 {
-				return nil, multierr.Append(splitErrors,
-					errors.Annotatef(berrors.ErrRestoreNoPeer, "region[%d] doesn't have any peer", regionInfo.Region.GetId()))
-			}
-			peer = regionInfo.Region.Peers[0]
+	var peer *metapb.Peer
+	// scanRegions may return empty Leader in https://github.com/tikv/pd/blob/v4.0.8/server/grpc_service.go#L524
+	// so wee also need check Leader.Id != 0
+	if regionInfo.Leader != nil && regionInfo.Leader.Id != 0 {
+		peer = regionInfo.Leader
+	} else {
+		if len(regionInfo.Region.Peers) == 0 {
+			return nil, multierr.Append(splitErrors,
+				errors.Annotatef(berrors.ErrRestoreNoPeer, "region[%d] doesn't have any peer", regionInfo.Region.GetId()))
 		}
+		peer = regionInfo.Region.Peers[0]
+	}
+	for i := 0; i < splitRegionMaxRetryTime; i++ {
```

Comment:
i don't think you should put those `peer = regionInfo.Leader` out of the retry loop, because in line 277–301 (`if resp.RegionError != nil`) the `regionInfo` is reassigned.

Reply:
you are right, I'll fix it.